File: ProjectTest/OneLang/Java/src/main/java/OneLang/One/Transforms/FillMutabilityInfo/FillMutabilityInfo.java
package OneLang.One.Transforms.FillMutabilityInfo;

import OneLang.One.AstTransformer.AstTransformer;
import OneLang.One.Ast.Types.IVariable;
import OneLang.One.Ast.Types.MutabilityInfo;
import OneLang.One.Ast.Expressions.Expression;
import OneLang.One.Ast.Expressions.BinaryExpression;
import OneLang.One.Ast.Expressions.InstanceMethodCallExpression;
import OneLang.One.Ast.References.VariableReference;
import OneLang.One.Ast.Statements.VariableDeclaration;

import OneLang.One.AstTransformer.AstTransformer;
import OneLang.One.Ast.Types.IVariable;
import OneLang.One.Ast.Types.MutabilityInfo;
import OneLang.One.Ast.References.VariableReference;
import OneLang.One.Ast.Statements.VariableDeclaration;
import OneStd.Objects;
import OneLang.One.Ast.Expressions.Expression;
import OneLang.One.Ast.Expressions.BinaryExpression;
import OneLang.One.Ast.Expressions.InstanceMethodCallExpression;

public class FillMutabilityInfo extends AstTransformer {
    public FillMutabilityInfo()
    {
        super("FillMutabilityInfo");
        
    }
    
    protected IVariable getVar(VariableReference varRef) {
        var v = varRef.getVariable();
        v.setMutability(v.getMutability() != null ? v.getMutability() : new MutabilityInfo(true, false, false));
        return v;
    }
    
    protected VariableReference visitVariableReference(VariableReference varRef) {
        this.getVar(varRef).getMutability().unused = false;
        return null;
    }
    
    protected VariableDeclaration visitVariableDeclaration(VariableDeclaration stmt) {
        super.visitVariableDeclaration(stmt);
        if (stmt.getAttributes() != null && Objects.equals(stmt.getAttributes().get("mutated"), "true"))
            stmt.getMutability().mutated = true;
        return null;
    }
    
    protected Expression visitExpression(Expression expr) {
        super.visitExpression(expr);
        
        if (expr instanceof BinaryExpression && ((BinaryExpression)expr).left instanceof VariableReference && Objects.equals(((BinaryExpression)expr).operator, "="))
            this.getVar(((VariableReference)((BinaryExpression)expr).left)).getMutability().reassigned = true;
        else if (expr instanceof InstanceMethodCallExpression && ((InstanceMethodCallExpression)expr).object instanceof VariableReference && ((InstanceMethodCallExpression)expr).getMethod().getAttributes().containsKey("mutates"))
            this.getVar(((VariableReference)((InstanceMethodCallExpression)expr).object)).getMutability().mutated = true;
        return null;
    }
    
    protected IVariable visitVariable(IVariable variable) {
        variable.setMutability(variable.getMutability() != null ? variable.getMutability() : new MutabilityInfo(true, false, false));
        return null;
    }
}

File: ProjectTest/OneLang/Python/OneLang/One/Transforms/FillMutabilityInfo.py
from OneLangStdLib import *
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.Statements as stats

class FillMutabilityInfo(astTrans.AstTransformer):
    def __init__(self):
        super().__init__("FillMutabilityInfo")
    
    def get_var(self, var_ref):
        v = var_ref.get_variable()
        v.mutability = v.mutability or types.MutabilityInfo(True, False, False)
        return v
    
    def visit_variable_reference(self, var_ref):
        self.get_var(var_ref).mutability.unused = False
        return None
    
    def visit_variable_declaration(self, stmt):
        super().visit_variable_declaration(stmt)
        if stmt.attributes != None and stmt.attributes.get("mutated") == "true":
            stmt.mutability.mutated = True
        return None
    
    def visit_expression(self, expr):
        super().visit_expression(expr)
        
        if isinstance(expr, exprs.BinaryExpression) and isinstance(expr.left, refs.VariableReference) and expr.operator == "=":
            self.get_var(expr.left).mutability.reassigned = True
        elif isinstance(expr, exprs.InstanceMethodCallExpression) and isinstance(expr.object, refs.VariableReference) and "mutates" in expr.method.attributes:
            self.get_var(expr.object).mutability.mutated = True
        return None
    
    def visit_variable(self, variable):
        variable.mutability = variable.mutability or types.MutabilityInfo(True, False, False)
        return None